zerver/data_import/rocketchat.py

```diff
@@ -0,0 +1,673 @@
+    user_id_to_recipient_id: Dict[int, int],
+    stream_id_mapper: IdMapper,
+    stream_id_to_recipient_id: Dict[int, int],
+    direct_id_to_direct_map: Dict[str, Dict[str, Any]],
+    dsc_id_to_dsc_map: Dict[str, Dict[str, Any]],
+    # zerver_realmemoji: List[Dict[str, Any]],
+    total_reactions: List[ZerverFieldsT],
+    output_dir: str,
+) -> None:
+    def list_reactions(reactions: Dict[str, Dict[str, Any]]) -> List[Dict[str, Any]]:
+        # List of dictionaries of form:
+        # {"name": "smile", "user_id": 2}
+        reactions_list: List[Dict[str, Any]] = []
+        for react_code in reactions:
+            name = react_code.split(":")[1]
+            usernames = reactions[react_code]["usernames"]
+
+            for username in usernames:
+                rc_user_id = username_to_user_id_map[username]
+                user_id = user_id_mapper.get(rc_user_id)
+                reactions_list.append({"name": name, "user_id": user_id})
+
+        return reactions_list
+
+    def message_to_dict(message: Dict[str, Any]) -> Dict[str, Any]:
+        rc_sender_id = message["u"]["_id"]
+        sender_id = user_id_mapper.get(rc_sender_id)
+        content = message["msg"]
+
+        if message.get("reactions"):
+            reactions = list_reactions(message["reactions"])
+        else:
+            reactions = []
+
+        message_dict = dict(
+            sender_id=sender_id,
+            content=content,
+            date_sent=int(message["ts"].timestamp()),
+            reactions=reactions,
+        )
+
+        # Add recipient_id and topic to message_dict
+        if is_pm_data:
+            direct_channel_id = message["rid"]
+            rc_member_ids = direct_id_to_direct_map[direct_channel_id]["uids"]
+            if rc_sender_id == rc_member_ids[0]:
+                zulip_member_id = user_id_mapper.get(rc_member_ids[1])
+                message_dict["recipient_id"] = user_id_to_recipient_id[zulip_member_id]
+            else:
+                zulip_member_id = user_id_mapper.get(rc_member_ids[0])
+                message_dict["recipient_id"] = user_id_to_recipient_id[zulip_member_id]
+            # PMs don't have topics, but topic_name field is required in `build_message`.
+            message_dict["topic_name"] = "Imported from rocketchat"
+        elif message["rid"] in dsc_id_to_dsc_map:
+            # Message is in a discussion
+            dsc_channel = dsc_id_to_dsc_map[message["rid"]]
+            parent_channel_id = dsc_channel["prid"]
+            stream_id = stream_id_mapper.get(parent_channel_id)
+            message_dict["recipient_id"] = stream_id_to_recipient_id[stream_id]
+            message_dict["topic_name"] = "(Discussion) {}".format(dsc_channel["fname"])
+        else:
+            stream_id = stream_id_mapper.get(message["rid"])
+            message_dict["recipient_id"] = stream_id_to_recipient_id[stream_id]
+            message_dict["topic_name"] = "Main channel content (imported from rocketchat)"
```

Comment:
And then this can be just "Imported from Rocket.Chat" to be consistent with our other import tools.